src/response.rs

```diff
@@ -90,6 +93,20 @@ pub fn send_redirect(route: String, location: String, status: u16) -> Result<Res
     Ok(bldr.body(None)?)
 }
 
+pub fn send_unauthorized(with_challenge: bool) -> Result<Response> {
+    if with_challenge {
+        let realm =
+            std::env::var(BASIC_AUTH_REALM).unwrap_or(FALLBACK_BASIC_AUTH_REALM.to_string());
+        let response_builder = Builder::new().status(401).header(
+            http::header::WWW_AUTHENTICATE,
+            format!("Basic realm=\"{}\"", realm),
+        );
+
+        return Ok(response_builder.body(None)?);
+    }
+
+    Ok(Builder::new().status(401).body(None)?)
```

Comment:
Can we add a body saying "Sign-in failed"?

Reply:
I could explicitly send a `WWW-Authenticate` header... let me see if the client picks that up and renders the contents

Reply:
I've updated the PR, it will now send the `www-authenticate` header for all 401, which makes the browser re-render the credential form. actual header values are not incorporated into the credential dialog itself.